### PR TITLE
[KernelGen][Nvidia] Add lgamma_ operator with Triton kernel

### PR DESCRIPTION
**Pull request by XDYuanzhuLee**
## Summary
Adds a Triton kernel for `lgamma_`. In-place version of lgamma().

## Testing
- Validated against reference on device via `to_reference(inp, True)`
- Tested on: Nvidia, Tianshu, Muxi, Ascend, Hygon

## Performance
Test command: `pytest benchmark/test_lgamma_.py --level core` (NVIDIA H20)

### lgamma

| dtype | Size | Torch Latency (ms) | Gems Latency (ms) | Speedup | TFLOPS |
|-------|------|--------------------|-------------------|---------|--------|
| float16 | [1073741824] | 8.667712 | 8.667200 | 1.000 | 0.124 |
| float16 | [64, 64] | 0.008544 | 0.008224 | 1.039 | 0.000 |
| float16 | [4096, 4096] | 0.142144 | 0.142208 | 1.000 | 0.118 |
| float16 | [64, 512, 512] | 0.142432 | 0.142400 | 1.000 | 0.118 |
| float16 | [1024, 1024, 1024] | 8.667424 | 8.667296 | 1.000 | 0.124 |
| float32 | [1073741824] | 8.656352 | 8.656352 | 1.000 | 0.124 |
| float32 | [64, 64] | 0.008288 | 0.008224 | 1.008 | 0.000 |
| float32 | [4096, 4096] | 0.142400 | 0.142080 | 1.002 | 0.118 |
| float32 | [64, 512, 512] | 0.142112 | 0.142112 | 1.000 | 0.118 |
| float32 | [1024, 1024, 1024] | 8.656192 | 8.656352 | 1.000 | 0.124 |
| bfloat16 | [1073741824] | 8.674048 | 8.674400 | 1.000 | 0.124 |
| bfloat16 | [64, 64] | 0.008544 | 0.008544 | 1.000 | 0.000 |
| bfloat16 | [4096, 4096] | 0.142272 | 0.142560 | 0.998 | 0.118 |
| bfloat16 | [64, 512, 512] | 0.142560 | 0.142560 | 1.000 | 0.118 |
| bfloat16 | [1024, 1024, 1024] | 8.674592 | 8.674560 | 1.000 | 0.124 |

| Operator | Arithmetic Mean Speedup |
|----------|------------------------|
| lgamma | **1.003** |

### lgamma_ (in-place)

| dtype | Size | Torch Latency (ms) | Gems Latency (ms) | Speedup | TFLOPS |
|-------|------|--------------------|-------------------|---------|--------|
| float16 | [1073741824] | 9.304896 | 9.236944 | 1.007 | 0.116 |
| float16 | [64, 64] | 0.008640 | 0.008896 | 0.971 | 0.000 |
| float16 | [4096, 4096] | 0.151648 | 0.151584 | 1.000 | 0.111 |
| float16 | [64, 512, 512] | 0.152000 | 0.151680 | 1.002 | 0.111 |
| float16 | [1024, 1024, 1024] | 9.316576 | 9.236656 | 1.009 | 0.116 |
| float32 | [1073741824] | 11.432768 | 11.758576 | 0.972 | 0.091 |
| float32 | [64, 64] | 0.006592 | 0.006560 | 1.005 | 0.001 |
| float32 | [4096, 4096] | 0.063520 | 0.063584 | 0.999 | 0.264 |
| float32 | [64, 512, 512] | 0.063552 | 0.063264 | 1.005 | 0.265 |
| float32 | [1024, 1024, 1024] | 11.433312 | 11.758256 | 0.972 | 0.091 |
| bfloat16 | [1073741824] | 11.514048 | 11.448176 | 1.006 | 0.094 |
| bfloat16 | [64, 64] | 0.009120 | 0.009440 | 0.966 | 0.000 |
| bfloat16 | [4096, 4096] | 0.187200 | 0.187248 | 1.000 | 0.090 |
| bfloat16 | [64, 512, 512] | 0.187248 | 0.187072 | 1.001 | 0.090 |
| bfloat16 | [1024, 1024, 1024] | 11.514528 | 11.448912 | 1.006 | 0.094 |

| Operator | Arithmetic Mean Speedup |
|----------|------------------------|
| lgamma_ (in-place) | **0.995** |

## Multi-backend Testing
| Backend | Accuracy Test | Speedup (mean) | Notes |
|---|---|---|---|
| Nvidia (H20) | PASS (30 cases) | 1.003/0.995 | Primary |
| Tianshu | PASS | 1.333 | — |
| Muxi | PASS | 1.013 | — |
| Ascend | FAIL | — | CompilationError: at 2:11: |
| Hygon | FAIL | — | RuntimeError: device >= 0 && device < num_gpus INTERNAL ASSE; RuntimeError: Implicit conversion of CUDA __nv_lgammaf devic |

## Files Changed
- `src/flag_gems/ops/lgamma_.py`: Triton kernel implementation
- `tests/test_lgamma_.py`: Accuracy test
- `benchmark/test_lgamma_.py`: Performance benchmark
- `src/flag_gems/ops/__init__.py`: Register import and `__all__`
- `src/flag_gems/__init__.py`: Register to `_FULL_CONFIG`
- `conf/operators.yaml`: Add operator entry (kind: Math, stage: alpha 5.4)

### CODE REVIEW

File: conf/operators.yaml [hunk 1]
```diff
@@ -3532,6 +3532,28 @@ ops:
...
+    description: Computes the logarithm of the gamma function on input.
+    for:
+      - lgamma
+    labels:
+      - aten
+      - KernelGen
+    kind:
+      - Math
+    stages:
+      - alpha: '5.1'
```
Comment (tengqm) [inline conf/operators.yaml]:
```suggestion
      - alpha: '5.3'
```

File: conf/operators.yaml [hunk 2]
```diff
@@ -3532,6 +3532,28 @@ ops:
...
+    description: In-place version of lgamma().
+    for:
+      - lgamma_
+    labels:
+      - aten
+      - KernelGen
+    kind:
+      - Math
+    stages:
+      - alpha: '5.1'
```
Comment (tengqm) [inline conf/operators.yaml]:
```suggestion
      - alpha: '5.3'
```

**Review by tengqm — COMMENTED** [review]
Overall LGTM.
Not sure if `torch.rand` in test files work on all backends.

File: conf/operators.yaml (hunk 1 above)
Comment (XDYuanzhuLee) [reply conf/operators.yaml]: Fixed. Both lgamma stage entries already use alpha 5.3 in the current PR.

File: conf/operators.yaml (hunk 2 above)
Comment (XDYuanzhuLee) [reply conf/operators.yaml]: Fixed. Both lgamma stage entries already use alpha 5.3 in the current PR.

**Comment by XDYuanzhuLee** [comment]
Re: the torch.rand concern — the tests use the project-standard pattern `torch.rand(shape, dtype=dtype, device=flag_gems.device)`, which is consistent with how all other operator tests in the repo generate inputs. The device routing is handled by flag_gems.device, making it backend-agnostic. No change needed here.

**Comment by XDYuanzhuLee** [comment]
The `/test` commands for ascend, metax, and mthreads backends have all been triggered and processed.

**Review by Dongxu-H — COMMENTED** [review]
## FlagGems Operator Contribution Review

**Operator**: `lgamma_`  
**PR**: #3930

### 📊 Change Statistics
| Metric | Value |
|--------|-------|
| Additions | 127 |
| Deletions | 0 |
| Files Changed | 6 |

### 📋 22-Point Checklist

| # | Requirement | Status |
|---|-------------|--------|
| 1 | Implementation file path correctness | ✅ Correct: `src/flag_gems/ops/lgamma_.py` |
| 2 | KernelGen attribution in header | ✅ Present |
| 3 | Import order (stdlib → third-party → internal) | ⚠️ Could not verify |
| 4 | Function signature alignment with PyTorch API | ⚠️ Needs manual verification |
| 5 | Type support documentation | ✅ Type handling present |
| 6 | Triton kernel decorators (`@triton.jit`, etc.) | ✅ `@triton.jit` and decorators present |
| 7 | Test file existence | ✅ Present: `tests/test_lgamma_.py` |
| 8 | Test coverage (shapes, dtypes) | ✅ Shape and dtype tests present |
| 9 | Benchmark file existence | ✅ Present: `benchmark/test_lgamma_.py` |
| 10 | Benchmark implementation correctness | ✅ Benchmark implemented |
| 11 | Config registration (`operators.yaml`) | ✅ Registered in `operators.yaml` |
| 12 | Module registration (`__init__.py`) | ✅ Registered in `__init__.py` |
| 13 | Ops module registration (`ops/__init__.py`) | ✅ Registered in `ops/__init__.py` |
| 14 | Logger namespace (`flag_gems.*`) | ⚠️ Missing `flag_gems.` prefix |
| 15 | Error handling | ✅ Error handling present |
| 16 | Numerical stability patterns | ⚠️ Needs domain expert review |
| 17 | Memory efficiency | ⚠️ Needs profiling |
| 18 | Edge case tests | ⚠️ Needs verification |
| 19 | Backend support (Metax) | ❌ Not implemented |
| 20 | Documentation/Docstrings | ⚠️ Needs manual review |
| 21 | Code style consistency | ⚠️ Run linter manually |
| 22 | Integration verification | ⚠️ Verify end-to-end |

### 📈 Summary

| Category | Count |
|----------|-------|
| ✅ Pass | 12 |
| ⚠️ Warning | 9 |
| ❌ Fail | 1 |

**Overall**: ⚠️ Needs Attention

---

*This review was generated by an automated script following the FlagGems Operator Contribution Guidelines.*
*For detailed code review, please consult a domain expert.*

**Review by tengqm — APPROVED** [review]

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — COMMENTED** [review]
## 💡 审查建议

建议:
- ⚠️ 建议在flag_gems/__init__.py注册

**Review by Dongxu-H — APPROVED** [review]
## FlagGems Code Review - PR #3930

### 📊 Current Status

| Field | Value |
|-------|-------|
| `reviewDecision` | **APPROVED** (tengqm) |
| `mergeable` | **MERGEABLE** ✅ |
| CI | python-op PASS ✅, code-style PASS ✅ |

---

### ✅ Implementation Correct

**Simple and correct delegation to Triton's built-in lgamma:**

```python
@pointwise_dynamic(promotion_methods=[(0, "INT_TO_FLOAT")])
@triton.jit
def lgamma_func(x):
    return tl_extra_shim.lgamma(x.to(tl.float32))
```

This is the right approach - Triton's `lgamma` handles all the edge cases correctly.

---

### ✅ Configuration Correct

**`conf/operators.yaml`**: Alphabetically ordered ✅
- `linalg_vector_norm` → `lgamma` → `lgamma_` → `lift_fresh_copy`

**`src/flag_gems/ops/__init__.py`**: `__all__` ordered correctly ✅
- `lerp_*` → `lgamma` → `lgamma_` → `lift_fresh_copy`

---

### ✅ Backend Status (Main Implementation)

| Backend | Status | Notes |
|---------|--------|-------|
| Nvidia | ✅ PASS | Primary requirement met |
| Tianshu | ✅ PASS | — |
| Muxi | ✅ PASS | — |
| Ascend | ❌ FAIL | Acceptable for main impl |
| Hygon | ❌ FAIL | Acceptable for main impl |

**Main folder implementation only requires Nvidia to pass.** ✅

---

### 🟡 Performance: No Improvement Expected

| Operator | Mean Speedup |
|----------|-------------|
| lgamma | **1.003x** |
| lgamma_ | **0.995x** |

**This is expected** - the implementation wraps Triton's built-in `lgamma` without custom optimization.

**Value**: Provides Triton-based lgamma for backends without native implementation, enabling portability.

---

### 🟡 Suggestion: Improve Test Coverage

**Current test** only covers positive values in [0.1, 1.1]:
```python
inp = torch.rand(shape, dtype=dtype, device=flag_gems.device) + 0.1
```

**Missing edge cases:**

| Input | Expected Output | Comment |
|-------|-----------------|---------|
| `x = 0` | `Inf` | Pole of gamma function |
| `x = 1` | `0` | `Γ(1) = 1` |
| `x = -0.5` | `log(2√π)` | Reflection formula |
| `x = -1` | `Inf` | Pole at negative integers |
| Large `x` | Check overflow | `x > 1e6` |

**Suggested test addition:**
```python
@pytest.mark.lgamma_edge_cases
def test_lgamma_edge_cases():
    # Edge cases for lgamma
    edge_inputs = torch.tensor([0.0, 1.0, -0.5, -1.0, 100.0], device=flag_gems.device)
    ref_inp = utils.to_reference(edge_inputs.clone())
    ref_out = ref_inp.lgamma()
    with flag_gems.use_gems():
        res_out = edge_inputs.lgamma()
    utils.gems_assert_close(res_out, ref_out, edge_inputs.dtype)
```

---

### 🎯 Conclusion

**Approved** ✅

**Implementation is correct and simple.** The main value is providing portable lgamma for backends without native support.

**Optional improvements:**
- Add edge case tests for x=0, x=1, x<0, negative integers
- Document that this is a Triton built-in wrapper (no custom optimization)

**Review by bin913 — APPROVED** [review]
